subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/ConfigurationCacheEnvironmentChangeTracker.kt

```diff
@@ -156,9 +192,14 @@ class EnvironmentChangeTracker(private val problemFactory: ProblemFactory) {
                 System.getProperties().clear()
             }
 
-            storedState.updates.forEach { update ->
-                System.getProperties()[update.key] = update.value
-            }
+            storedState.updates
+                // If property was overridden after storing we don't want to load it with outdated value
+                .filter { update ->
+                    !overriddenSystemProperties.containsKey(update.key)
```

Comment:
I strongly suspect that the `SystemPropertyMutate` should be left intact, as such value indicates the system property overwritten in the build logic. If the last write was from build logic, then `-D` cannot overwrite it. Unless an included build causes start parameter flags to be reapplied :sob:. But I suggest addressing the latter separately.

```suggestion
                    (update.value is SystemPropertyLoad) && !overriddenSystemProperties.containsKey(update.key)
```

I don't think we have a test for this, as most ways to update system properties are also making them inputs. Could you please add one to `UndeclaredBuildInputsIntegrationTest`? You can use `System.getProperties().putAll(...)` to do the update without adding inputs.

Reply:
Actually, after looking at the test failures, it looks like we need to be even more careful when considering removals and cleanups here.

Reply:
But correct condition is `update !is SystemPropertyLoad || !overriddenSystemProperties.contains(update.key)` I guess?

Reply:
Indeed, it is.